### PR TITLE
#3870 sorted set

### PR DESCRIPTION
**Pull request by tucek**
## Basics

These points need to be fulfilled for every PR:

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [x] add a line in `doc/news/_preparation_next_release.md`
  - [x] reformat the code with `scripts/dev/reformat-all`
  - [x] make all unit tests pass
  - [x] fix all memleaks
- [x] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [x] I added unit tests for my code
- [x] I fully described what my PR does in the documentation
      (not in the PR description)
- [x] I fixed all affected documentation
- [x] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [x] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [x] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

**Comment by tucek** [comment]
This PR is based on #3991

**Comment by tucek** [comment]
I've hit a little bit of a roadblock because of the contract for `SortedSet::subSet`, `SortedSet::headSet` and `SortedSet::tailSet`
(https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/SortedSet.html).
These 3 methods have to return a view of the original set, reflecting changes done to the original set in the view and vice versa.

@kodebach any ideas on that? The only obvious solution would introduce a lot of logic to the Java `KeySet` to register and update views created by this 3 methods. It also would hold references to these "view" keysets without knowing when they are not longer needed, prventing gc and therefore leading to memory leaks...

**Comment by kodebach** [comment]
Since the `KeySet` class is just a "view" of the C implementation anyway, you could just add `private Key start` and `private Key end` fields to the `KeySet` class that are normally null, but `subSet`, `headSet` and `tailSet` set these fields appropriately.

```java
@Override
public KeySet tailSet(Key toElement) {
    var ks = new KeySet (pointer);
    ks.end = toElement;
    return ks;
}
```

You would still need the bounds checks, but you don't need another class.

**Comment by tucek** [comment]
> Since the `KeySet` class is just a "view" of the C implementation anyway, you could just add `private Key start` and `private Key end` fields to the `KeySet` class that are normally null, but `subSet`, `headSet` and `tailSet` set these fields appropriately.
> 
> ```java
> @Override
> public KeySet tailSet(Key toElement) {
>     var ks = new KeySet (pointer);
>     ks.end = toElement;
>     return ks;
> }
> ```
> 
> You would still need the bounds checks, but you don't need another class.

Currently `KeySet` calls ksDel on clean-up. Therefore, IMHO this is not possible without key set reference counter... I would have to manage the _one_ KeySet representation in Java by introducing a kind of KeySet wrapper class ;(

**Comment by tucek** [comment]
Rebased branch with current master.
Please review. Thank you!

**Comment by tucek** [comment]
Unrelated fail of `FULL` check:
```
You have 1 outdated cask installed.
You can upgrade it with brew upgrade
or list it with brew outdated.
```

**Comment by tucek** [comment]
Rebased with current Elektra master.

File: src/bindings/jna/gradle/wrapper/gradle-wrapper.properties
Comment (markus2330) [inline src/bindings/jna/gradle/wrapper/gradle-wrapper.properties:3]: Please avoid unrelated changes in PRs.

File: src/bindings/jna/gradlew
Comment (markus2330) [inline src/bindings/jna/gradlew:1]: Please avoid unrelated changes in PRs.

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/Elektra.java
Comment (markus2330) [inline src/bindings/jna/libelektra/src/main/java/org/libelektra/Elektra.java:395]: This is not public API and as such it should not be used.

But probably it makes sense to provide `ksSearch`. The return type needs to be fixed though (elektraCursor).

File: doc/news/_preparation_next_release.md
Comment (markus2330) [inline doc/news/_preparation_next_release.md:160]: This is an extremely short description: No link to your source nor to SortedSet API docu. It does not say which benefits there are for the user. An updated tutorial would also be useful.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you for the PR! Please update docu and please split unrelated things to other PRs.

File: src/bindings/jna/gradle/wrapper/gradle-wrapper.properties
Comment (tucek) [reply src/bindings/jna/gradle/wrapper/gradle-wrapper.properties:3]: I've created #4037 and will remove this change from this PR

File: src/bindings/jna/gradlew
Comment (tucek) [reply src/bindings/jna/gradlew:1]: I've created #4037 and will remove this change from this PR

**Comment by tucek** [comment]
I've converted this PR back to draft state, because it is blocked by #4026

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/Elektra.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/Elektra.java:395]: tracked via #4026 and implemented in this PR.

**Comment by tucek** [comment]
I'm always getting the following build failure, when building with documentation (also tried it on current master):
```sh
[  0%] Generating ../../../doc/man/man7/elektra-related.7
CMake Error at scripts/cmake/ElektraManPage.cmake:21 (message):
  /usr/lib/ruby/vendor_ruby/ronn/roff.rb:354:in `gsub': incompatible encoding
  regexp match (Windows-31J regexp with UTF-8 string)
  (Encoding::CompatibilityError)

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:354:in `write'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:247:in `inline_filter'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:243:in `block in
  inline_filter'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:238:in `block in
  each'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:237:in `upto'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:237:in `each'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:243:in `inline_filter'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:120:in `block_filter'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:88:in `block in block_filter'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:238:in `block in
  each'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:237:in `upto'

        from /usr/lib/ruby/vendor_ruby/nokogiri/xml/node_set.rb:237:in `each'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:88:in `block_filter'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:91:in `block_filter'

        from /usr/lib/ruby/vendor_ruby/ronn/roff.rb:18:in `initialize'

        from /usr/lib/ruby/vendor_ruby/ronn/document.rb:245:in `new'

        from /usr/lib/ruby/vendor_ruby/ronn/document.rb:245:in `to_roff'

        from /usr/lib/ruby/vendor_ruby/ronn/document.rb:240:in `convert'

        from /usr/bin/ronn:220:in `block (2 levels) in <main>'

        from /usr/bin/ronn:199:in `each'

        from /usr/bin/ronn:199:in `block in <main>'

        from /usr/bin/ronn:184:in `each'

        from /usr/bin/ronn:184:in `<main>'



make[2]: *** [doc/help/CMakeFiles/man-elektra-related.dir/build.make:61: ../doc/man/man7/elektra-related.7] Error 1
make[2]: *** Deleting file '../doc/man/man7/elektra-related.7'
make[1]: *** [CMakeFiles/Makefile2:4766: doc/help/CMakeFiles/man-elektra-related.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

Has anybody any ideas?

File: doc/news/_preparation_next_release.md
Comment (tucek) [reply doc/news/_preparation_next_release.md]: I've updated the release notes

**Comment by tucek** [comment]
This PR now is ready for review again.

**Comment by kodebach** [comment]
> I'm always getting the following build failure, when building with documentation (also tried it on current master):

Do you have the correct version of ronn? I think we switched at some point. AFAIK you need this version: https://github.com/apjanke/ronn-ng/releases/tag/v0.10.0-beta1

**Comment by tucek** [comment]
> > I'm always getting the following build failure, when building with documentation (also tried it on current master):
> 
> Do you have the correct version of ronn? I think we switched at some point. AFAIK you need this version: https://github.com/apjanke/ronn-ng/releases/tag/v0.10.0-beta1

Thank you! Updating from 0.8.0 to Ronn-NG v0.9.1 did work.

File: src/libs/elektra/keyset.c
Comment (markus2330) [inline src/libs/elektra/keyset.c:725]: we assert that:

```suggestion
	// The key already exist in kst.
```

File: src/libs/elektra/keyset.c
Comment (markus2330) [inline src/libs/elektra/keyset.c:728]:
```suggestion
	// The key does not exist in ks.
```

File: src/libs/elektra/keyset.c
Comment (markus2330) [inline src/libs/elektra/keyset.c:737]: We also need `@see` to ksLookup.

@lawli3t can you please also do a review of this docu here.

File: src/libs/elektra/keyset.c
Comment (markus2330) [inline src/libs/elektra/keyset.c:712]: As discussed, ksSearchInternal should be static.

File: src/libs/elektra/keyset.c
Comment (markus2330) [inline src/libs/elektra/keyset.c:713]: Better to not promise something we don't do:

```suggestion
 *  Search key in a key set, either yielding the actual index
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thank you for updating the PR!

**Comment by markus2330** [comment]
There are some small open review comments, otherwise this PR would be fine. Please fix to get this merged.

**Comment by tucek** [comment]
Fixed corrupted branch

**Comment by markus2330** [comment]
Please readd "ready to merge" if everything is done.

File: src/libs/elektra/keyset.c
Comment (tucek) [reply src/libs/elektra/keyset.c:725]: This was inherited from `ksSearchInternal` and has been updated in both function docs.

File: src/libs/elektra/keyset.c
Comment (tucek) [reply src/libs/elektra/keyset.c:728]: This was inherited from `ksSearchInternal` and has been updated in both function docs.

File: src/libs/elektra/keyset.c
Comment (tucek) [reply src/libs/elektra/keyset.c:737]: Updated ksSearch docs with `@see ksLookup() for retrieving the found key`

File: src/libs/elektra/keyset.c
Comment (tucek) [reply src/libs/elektra/keyset.c:713]: I thought it would be the intended way this function works after the hash map lookup is being implemented. Removed `Efficient`

File: src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetException.java
Comment (tucek) [reply src/bindings/jna/libelektra/src/main/java/org/libelektra/exception/KeySetException.java:6]: I have added the following paragraph to `KeyException` and `KeySetException`:

This exception is related to unrecoverable C API specific errors (primarily allocation problems). All other exceptional states are represented by more specific exceptions. For more detailed information see exceptions in this package as well as a method's individual documentation.

File: src/libs/elektra/keyset.c
Comment (tucek) [reply src/libs/elektra/keyset.c:711]: done

**Comment by tucek** [comment]
Rdy for merge. All CI failures are unrelated.

**Review by markus2330 — APPROVED** [review]
Thank you, especially for also adding ksSearch! :sparkling_heart:

**Comment by tucek** [comment]
Branch has been rebased to current master

**Review by mpranj — APPROVED** [review]
Thank you, great work! 👍

File: doc/news/_preparation_next_release.md
Comment (mpranj) [inline doc/news/_preparation_next_release.md:175]:
```suggestion
```

**Pull request merged**
